### DISCUSSION TITLE
How to Generate Animated Webp from a GIF image

### DISCUSSION THREAD

**Discussion by AmitSonkhiya** [libvips/php-vips#210] [Q&A] [answered]
I'm trying to figure out generating animated webp from a GIF image.
My purpose is reducing file size (from 2-10 MB to comparatively lesser).

From this [Stackoverflow link](https://stackoverflow.com/questions/63003350/animated-not-working-when-i-convert-gif-to-webp-in-php), I have tried as follow:
```
$image = Vips\Image::newFromFile('something.gif', [
    'n' => -1,
    'access' => 'sequential'
]); 
$image->writeToFile('something.webp');
```

However, that is causing the following error:

> optional argument 'n' doesn't exist.

From the Libvips documentation, I think that the argument ``'n' has removed.
So after removing the argument  `'n' => -1`, the webp image is generated but it is loosing the animation.

What is the proper code to generate animated webp from an animated GIF image with lesser file size?

Thank you

PHP version: 8.2.5
Libvips version: 8.14.2
PHP vips version: 2.1.1
OS: Mac OS
GIF size: 1.94 MB
Generated webp size (without animation): 25

**Comment by jcupitt** [comment]
Hi @AmitSonkhiya, it should work, I guess your libvips has been built without a GIF loader:

https://www.libvips.org/API/current/VipsForeignSave.html#vips-gifload

**Reply by AmitSonkhiya** [reply]
Hi,
Thank you for quick response.

How do I verify if the `gifload` is absent?
I have run `vips gifload` on CLI and getting usages description, through I'm not sure it is the correct way to verify.

Thank you

**Reply by jcupitt** [reply, marked as answer]
Could you have two libvipses installed? I tried this:

```php
#!/usr/bin/env php
<?php
require __DIR__ . '/vendor/autoload.php';
use Jcupitt\Vips;

# Vips\Config::setLogger(new Vips\DebugLogger());
$version =  Vips\Config::version();
echo "libvips version $version\n";

$image = Vips\Image::newFromFile("/Users/john/pics/3198.gif", [
    "n" => -1,
    "access" => "sequential"
]);
$image->writeToFile("x.webp");
```

And it works fine on my mac. That's with libvips from homebrew.

**Reply by jcupitt** [reply]
ie. perhaps your php is picking up a very old version or different version libvips.

**Reply by AmitSonkhiya** [reply]
Hi,

libsvips version 8.14.2

Thank you for the support.
Your example works fine and I'm getting the animated webp.
I was simply appending .webp to original GIF name as target file (something.gif => something.gif.webp).
I guess here was the issue. After giving another random name, it worked perfectly.

Just a quick question more.
How could additionally reduce the target webp file size?
GIF size: 1.94MB
Animated webp size: 420KB

Thank you

**Reply by jcupitt** [reply]
You can turn down Q to get more highly compressed images:

```
$image->writeToFile("x.webp", [
    "Q" => 30
]);
```

Though of course quality may become worse, and the improvement will depend on the image contents.

**Reply by jcupitt** [reply]
> I was simply appending .webp to original GIF name as target file (something.gif => something.gif.webp). I guess here was the issue. After giving another random name, it worked perfectly.

That should be fine, you probably had some other minor bug.

**Reply by AmitSonkhiya** [reply]
Yeah! It needs an attention to discover the bug, may be at the framework level.

Thanks for all the guidance and prompt response.
Have a good time ahead.

**Comment by AmitSonkhiya** [comment]
Hi @jcupitt ,

I'm opening the discussion again as I'm stuck here.

With the `Image::thumbnail()` method, I'm able to create thumbnails of given sizes.
However, as obvious, thumbnails are losing the animation.

I have tried the resize method but that requires a scale parameter rather than specific size (width).
How do I create thumbnails that preserve the animation (GIF and Webp)?
I could not find any purposeful argument :-/

```
(Image::thumbnail('image.webp', 384))->writeToFile('image_384.jpg');
```

Thank you again

**Reply by jcupitt** [reply]
Just use `n=-1`, meaning all frames.

```php
Image::thumbnail('image.webp[n=-1]', 384)->writeToFile('image_384.jpg');
```

**Reply by AmitSonkhiya** [reply]
@jcupitt 

Thank again for your time and I would need your guidance a bit more.
I'm trying to create animated thumbnail from animated image.

### 1. File does not exist for `thumbnail()`

I have followed your previous advice, however the code obviously gave me the error:
I have the file `image.webp` but it is how programs treat file name.

> libvips error: VipsForeignLoad: file "image.webp[n=-1]" does not exist

```
$image = Image::thumbnail('image.webp[n=-1]', 512);
$image->writeToFile('target.webp')
```

### 2. Produced animated thumbnail using `thumbnail_image()`
The following code produced animated thumbnail.

```
$image = Image::newFromFile('source.webp', ["n" => -1, "access" => "sequential"]);
$thumbnail = $image->thumbnail_image(512);
$thumbnail->writeToFile('target.webp');
```
But as [documented in Libvips](https://www.libvips.org/API/current/libvips-resample.html#vips-thumbnail-image), this operation is slower and produces poor quality output, and it was just to test so _I would rather stick with your solution._

### 3. Generating thumbnails using width side only

It seems that both `thumbnail()` and `thumbnail_image()` take higher side into consideration while generating the thumbnail.
So if image is taller then the height will be reduced to the given parameter and width will be proportional to it.

**How do I force PHP Vips to consider width only in the account and produce height proportional to it? Similar to [this link](https://github.com/libvips/libvips/issues/781)?**

There is the following error while trying the `size` argument.

> libvips error: php-vips: enum 'VipsSize' has no member '512x', should be one of: both, up, down, force

```
Image::thumbnail('source.jpg', 512, ['size' => '512x']);
```

Thank you

**Reply by jcupitt** [reply]
I works for me. With this program:

```php
#!/usr/bin/env php
<?php

require __DIR__ . '/vendor/autoload.php';
use Jcupitt\Vips;

Vips\Image::thumbnail('dancing_banana2.lossless.webp[n=-1]', 100)->writeToFile('x.gif');
```

I see:

```
$ ./thumbnail.php 
$ 
```


To make:

![x](https://github.com/libvips/php-vips/assets/580843/2f27aecf-df80-4f83-913d-f44d7605525e)

**Reply by jcupitt** [reply]
Perhaps your libvips has been built without webp support?

**Reply by jcupitt** [reply]
> How do I force PHP Vips to consider width only in the account and produce height proportional to it?

Just use a huge value for the height parameter, eg. 10000000.

**Reply by AmitSonkhiya** [reply]
@jcupitt 

Thank you for the software and much appreciated support.
I had a slash missing in the path.

Fixing it made everything to work smoothly.

PS: I understand there are pretty efforts behind the library. Still having FAQs/TIPs section somewhere would be effective?
For example, as you told `image.gif[n=-1]` for animated thumbnails or load all frames.
Or setting up height to a big number for width constraint. 
Through, official docs have their own stand, just searching them with human keywords would be a breeze.
A Google search leads no where.
It will also helpful for you to answer basic questions like I asked through a link to them.
And it will attract more developers.

Thank you

**Reply by jcupitt** [reply]
I tried to improve the thumbnail docs a bit: https://github.com/libvips/libvips/commit/ce99675cc75a4f74658e746d241f938a388cba73

**Reply by AmitSonkhiya** [reply]
It is indeed better. 😊

**Comment by AmitSonkhiya** [comment]
I'm still digging more. Uhm!

@jcupitt I'm able to create SQUARE static thumbnails using the following code.

```
Vips\Image::thumbnail('source.jpg'), 384, ['height' => 384, 'crop' => 'attention'])
        ->writeToFile('target.jpg', ['strip' => 'strip']);
```

I'm curious if we can generate a square thumbnail from a rectangle image while preserving the animation?
And if yes then how and if it would be memory-extensive due to number of frames?
Of course, `source.gif[n=-1]` in the code above generates thumbnails but loses animation if the aspect ratio is changed.

Thank you

**Reply by jcupitt** [reply]
There's an enhancement issue for adding animated thumbnail crop https://github.com/libvips/libvips/issues/2668

But no one's done the work yet.

**Reply by AmitSonkhiya** [reply]
I see, no option other than waiting until an expert incorporate this enhancement.
